### PR TITLE
Feature/predefined formula

### LINKED ISSUE #316: Common pool autoscale forumla
Leverage the `BatchLabs-data` repo to have a list of premade forumlas that would be usefull for the user to get started

### PR DESCRIPTION
**Pull request by johnnyzhang82**
Fixed #316
* Added a list of predefined autoscale forrmulas.
* Fixed some keyword highlighting issues in editor.

### CODE REVIEW

File: app/components/pool/action/scale/autoscale-formula-picker.component.ts [hunk 1]
```diff
@@ -47,6 +51,8 @@ export class AutoscaleFormulaPickerComponent implements OnInit, OnDestroy, Contr
         this._subs.push(this.autoscaleFormulaService.formulas.subscribe((formulas) => {
             this.savedAutoscaleFormulas = formulas;
         }));
+        this.predefinedFormulaService.init();
```
Comment (timotheeguerin) [inline app/components/pool/action/scale/autoscale-formula-picker.component.ts]: I would move this to the `app.component` That's static enough so you don't want to reload every time

File: app/services/predefined-formula.service.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    public init() {
+        this.githubData.get(`${predefinedFormulaPath}${indexFile}`).subscribe({
+            next: (response: Response) => {
+                const result: AutoscaleFormula[] = new Array();
+                const data: SampleFormulaIndex[] = response.json();
+                for (let file of data) {
+                    this.githubData.get(`${predefinedFormulaPath}${file.value}`).subscribe({
+                        next: (fileResponse: Response) => {
+                            const text: string = fileResponse.text();
+                            const formula = <AutoscaleFormula>{
```
Comment (timotheeguerin) [inline app/services/predefined-formula.service.ts]: use `const forumla: AutoScaleForumla` instead of casting

File: app/utils/autoscale.ts [hunk 3]
```diff
@@ -17,21 +17,50 @@ const systemFunc = [
...
+    "TimeInterval_Week",
+    "TimeInterval_Year",
+];
+
 // tslint:disable-next-line:max-line-length
-const keywordRegex = /\$CPUPercent|\$WallClockSeconds|\$MemoryBytes|\$DiskBytes|\$DiskReadBytes\$DiskWriteBytes|\$DiskReadOps|\$DiskWriteOps|\$NetworkInBytes|\$NetworkOutBytes|\$SampleNodeCount|\$ActiveTasks|\$RunningTasks|\$PendingTasks|\$SucceededTasks|\$FailedTasks|\$CurrentDedicated|\$TargetDedicated/;
-const functionsRegex = /avg|len|lg|ln|log|max|min|norm|percentile|rand|range|std|stop|sum|time|val/;
-const obtainRegex = /GetSample|GetSamplePeriod|Count|HistoryBeginTime|GetSamplePercent/;
-const atomRegex = /true|false|null/;
+const keywordRegex = /\$\b(CPUPercent|WallClockSeconds|MemoryBytes|DiskBytes|DiskReadBytes\$DiskWriteBytes|DiskReadOps|DiskWriteOps|NetworkInBytes|NetworkOutBytes|SampleNodeCount|ActiveTasks|RunningTasks|PendingTasks|SucceededTasks|FailedTasks|CurrentDedicated|TargetDedicated|NodeDeallocationOption)\b/;
```
Comment (timotheeguerin) [inline app/utils/autoscale.ts]: Can't you do that for all the following?

```
const keywordRegex = new Regex(variables.join("|"))?
```

you might need to escape `$` too

File: app/components/pool/action/scale/autoscale-formula-picker.component.ts (hunk 1 above)
Comment (johnnyzhang82) [inline app/components/pool/action/scale/autoscale-formula-picker.component.ts]: Updated.

File: app/services/predefined-formula.service.ts (hunk 2 above)
Comment (johnnyzhang82) [inline app/services/predefined-formula.service.ts]: Updated.

File: app/utils/autoscale.ts (hunk 3 above)
Comment (johnnyzhang82) [inline app/utils/autoscale.ts]: Updated.

File: app/services/index.ts [hunk 4]
```diff
@@ -23,6 +23,7 @@ export * from "./vm-size.service";
 export * from "./adal";
 export * from "./batch-client.service";
 export * from "./storage-client.service";
+export * from "./predefined-formula.service";
```
Comment (ascobie) [inline app/services/index.ts]: not a biggie, but i like to keep these in alpha order where possible.

File: app/styles/pool/create.scss [hunk 5]
```diff
@@ -174,12 +174,32 @@ bl-vm-size-picker {
 
 bl-autoscale-formula-picker {
     display: flex;
+    .CodeMirror {
```
Comment (ascobie) [inline app/styles/pool/create.scss]: lowerCase? .codeMirror

File: app/services/index.ts (hunk 4 above)
Comment (timotheeguerin) [inline app/services/index.ts]: yeah `tslint` doesn't check for the alpha ordering in`index.ts` files, not sure why

File: app/styles/pool/create.scss (hunk 5 above)
Comment (timotheeguerin) [inline app/styles/pool/create.scss]: actually prefer `kebab-case` for css so `code-mirror`

Comment (johnnyzhang82) [inline app/styles/pool/create.scss]: This is a class defined by codemirror and override font size here.

File: app/services/index.ts (hunk 4 above)
Comment (johnnyzhang82) [inline app/services/index.ts]: tslint checks import sort, but not for export, I guess there are some dependencies?

File: app/services/predefined-formula.service.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+interface SampleFormulaIndex {
+    name: string;
+    value: string;
+}
+
+@Injectable()
+export class PredefinedFormulaService {
+    public predefinedFormulas: Observable<AutoscaleFormula[]>;
+    private _predefinedFormulas = new BehaviorSubject<AutoscaleFormula[]>(null);
+    constructor(private githubData: GithubDataService) {
```
Comment (timotheeguerin) [inline app/services/predefined-formula.service.ts]: add blank line before constructor

File: app/services/predefined-formula.service.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        this.predefinedFormulas = this._predefinedFormulas.asObservable();
+    }
+
+    public init() {
+        this.githubData.get(`${predefinedFormulaPath}${indexFile}`).subscribe({
+            next: (response: Response) => {
+                const result: AutoscaleFormula[] = new Array();
+                const data: SampleFormulaIndex[] = response.json();
+                for (let file of data) {
+                    this.githubData.get(`${predefinedFormulaPath}${file.value}`).subscribe({
```
Comment (timotheeguerin) [inline app/services/predefined-formula.service.ts]: can you move this nested one in another function so it looks cleaner?

**Pull request merged**